sources/file.go

```diff
@@ -0,0 +1,247 @@
+			if err != nil {
+				logging.Error().Str("path", s.FullPath()).Msg("could not create tmp file")
+				return nil
+			}
+			defer func() {
+				_ = tmpfile.Close()
+				_ = os.Remove(tmpfile.Name())
+			}()
+
+			_, err = io.Copy(tmpfile, reader)
+			if err != nil {
+				logging.Error().Str("path", s.FullPath()).Msg("could not copy archive file")
+				return nil
+			}
+
+			reader = tmpfile
+		}
+	}
+
+	return extractor.Extract(ctx, reader, func(_ context.Context, d archives.FileInfo) error {
+		if d.IsDir() {
+			return nil
+		}
+
+		innerReader, err := d.Open()
+		if err != nil {
+			logging.Error().Str("path", s.FullPath()).Err(err).Msg("could not open archive inner file")
+			return nil
+		}
+		path := filepath.Clean(d.NameInArchive)
+
+		if s.Config != nil && shouldSkipPath(s.Config, path) {
+			logging.Debug().Str("path", s.FullPath()).Msg("skipping file: global allowlist")
+			return nil
+		}
+
+		file := &File{
+			Content:         innerReader,
+			Path:            path,
+			Symlink:         s.Symlink,
+			outerPaths:      append(s.outerPaths, filepath.ToSlash(s.Path)),
+			MaxArchiveDepth: s.MaxArchiveDepth,
+			archiveDepth:    s.archiveDepth + 1,
+		}
+
+		if err := file.Fragments(ctx, yield); err != nil {
+			_ = innerReader.Close()
+			return err
+		}
+
+		_ = innerReader.Close()
+		return nil
+	})
+}
+
+// decompressorFragments recursively crawls archives and yields fragments
+func (s *File) decompressorFragments(decompressor archives.Decompressor, reader io.Reader, yield FragmentsFunc) error {
+	innerReader, err := decompressor.OpenReader(reader)
+	if err != nil {
+		logging.Error().Str("path", s.FullPath()).Msg("could read compressed file")
+		return nil
+	}
+
+	if err := s.fileFragments(bufio.NewReader(innerReader), yield); err != nil {
+		_ = innerReader.Close()
+		return err
+	}
+
+	_ = innerReader.Close()
+	return nil
+}
+
+// fileFragments reads the file into fragments to yield
+func (s *File) fileFragments(reader *bufio.Reader, yield FragmentsFunc) error {
+	// Create a buffer if the caller hasn't provided one
+	if s.Buffer == nil {
+		s.Buffer = make([]byte, defaultBufferSize)
+	}
+
+	totalLines := 0
+	for {
+		fragment := Fragment{
+			FilePath: s.FullPath(),
+		}
+
+		n, err := reader.Read(s.Buffer)
+		if n == 0 {
+			if err != nil && err != io.EOF {
+				return yield(fragment, fmt.Errorf("could not read file: %w", err))
+			}
+
+			return nil
+		}
+
+		// Only check the filetype at the start of file.
+		if totalLines == 0 {
+			// TODO: could other optimizations be introduced here?
+			if mimetype, err := filetype.Match(s.Buffer[:n]); err != nil {
+				return yield(
+					fragment,
+					fmt.Errorf("could not read file: could not determine type: %w", err),
+				)
+			} else if mimetype.MIME.Type == "application" {
+				logging.Debug().Str(
+					"mime_type", mimetype.MIME.Value,
+				).Str(
+					"path", s.FullPath(),
+				).Msgf(
+					"skipping binary file",
+				)
```

Comment:
This is kinda whack, why not something like?
```suggestion
				logging.Debug().
				    Str("mime_type", mimetype.MIME.Value).
				    Str("path", s.FullPath()).
				    Msgf("skipping binary file")
```

Reply:
Habit from Python (I prefer your version). :+1: